Fix exception raised when load an certificate with UTF-8 characters (…

paulosgf opened the pull request:
Fix exception raised when load an certificate with UTF-8 characters.

paulosgf commented:
Patch to fix a exception raised when the
 Openssl.crypto.load_certificate opens a certificate that contains UTF-8
 characters on the comments, and then tries to convert to ASCII before
 create an X509 object.

alex commented on src/OpenSSL/crypto.py:
```diff
@@ -1935,7 +1935,7 @@ def load_certificate(type, buffer):
     """
     if isinstance(buffer, _text_type):
-        buffer = buffer.encode("ascii")
+        buffer = buffer.encode("ascii", "ignore")
```
I don't think ignore makes sense here, can we just encode to utf-8?

alex commented on tests/test_debug.py:
```diff
@@ -1,10 +1,57 @@
+    """
+    Test get_cert() function and raises an error in case of UTF8 fixture
+    """
+    assert get_cert() is None
```
These should be in `test_crypto.py`, they're not a test of the debug interface.

alex commented on tests/test_debug.py:
```diff
@@ -1,10 +1,57 @@
+    print(x509.get_subject())
+
+
+def test_get_cert():
```
This isn't correctly named -- the point of this test isn't to test the `get_cert` function, it's to test non ascii PEM.

paulosgf commented:
Yes, it's makes sense. I'll change this soon and resubmit.

Em sex, 26 de mar de 2021 18:40, Alex Gaynor ***@***.***>
escreveu:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/OpenSSL/crypto.py
> <https://github.com/pyca/pyopenssl/pull/1007#discussion_r602599139>:
>
> > @@ -1935,7 +1935,7 @@ def load_certificate(type, buffer):
>      :return: The X509 object
>      """
>      if isinstance(buffer, _text_type):
> -        buffer = buffer.encode("ascii")
> +        buffer = buffer.encode("ascii", "ignore")
>
> I don't think ignore makes sense here, can we just encode to utf-8?
> ------------------------------
>
> In tests/test_debug.py
> <https://github.com/pyca/pyopenssl/pull/1007#discussion_r602599252>:
>
> > +
> +def get_cert():
> +    """
> +    Read a certificate, create a x509 object and print his subject
> +    """
> +    x509 = OpenSSL.crypto.load_certificate(
> +        OpenSSL.crypto.FILETYPE_PEM, cert_utf8
> +    )
> +    print(x509.get_subject())
> +
> +
> +def test_get_cert():
> +    """
> +    Test get_cert() function and raises an error in case of UTF8 fixture
> +    """
> +    assert get_cert() is None
>
> These should be in test_crypto.py, they're not a test of the debug
> interface.
> ------------------------------
>
> In tests/test_debug.py
> <https://github.com/pyca/pyopenssl/pull/1007#discussion_r602599490>:
>
> >
>  def test_debug_info():
>      """
>      Debug info contains correct data.
>      """
>      # Just check a sample we control.
>      assert version.__version__ in _env_info
> +
> +
> +def get_cert():
> +    """
> +    Read a certificate, create a x509 object and print his subject
> +    """
> +    x509 = OpenSSL.crypto.load_certificate(
> +        OpenSSL.crypto.FILETYPE_PEM, cert_utf8
> +    )
> +    print(x509.get_subject())
> +
> +
> +def test_get_cert():
>
> This isn't correctly named -- the point of this test isn't to test the
> get_cert function, it's to test non ascii PEM.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pyca/pyopenssl/pull/1007#pullrequestreview-622550658>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF7Q4S5DXP5YCMUPNT6T4FTTFT5NFANCNFSM4Z36IYEQ>
> .
>

alex commented:
Please update this PR, don't close and open a new one.

On Fri, Mar 26, 2021 at 9:58 PM Paulo S. G. Ferraz ***@***.***>
wrote:

> Yes, it's makes sense. I'll change this soon and resubmit.
>
> Em sex, 26 de mar de 2021 18:40, Alex Gaynor ***@***.***>
> escreveu:
>
> > ***@***.**** commented on this pull request.
> > ------------------------------
> >
> > In src/OpenSSL/crypto.py
> > <https://github.com/pyca/pyopenssl/pull/1007#discussion_r602599139>:
> >
> > > @@ -1935,7 +1935,7 @@ def load_certificate(type, buffer):
> > :return: The X509 object
> > """
> > if isinstance(buffer, _text_type):
> > - buffer = buffer.encode("ascii")
> > + buffer = buffer.encode("ascii", "ignore")
> >
> > I don't think ignore makes sense here, can we just encode to utf-8?
> > ------------------------------
> >
> > In tests/test_debug.py
> > <https://github.com/pyca/pyopenssl/pull/1007#discussion_r602599252>:
> >
> > > +
> > +def get_cert():
> > + """
> > + Read a certificate, create a x509 object and print his subject
> > + """
> > + x509 = OpenSSL.crypto.load_certificate(
> > + OpenSSL.crypto.FILETYPE_PEM, cert_utf8
> > + )
> > + print(x509.get_subject())
> > +
> > +
> > +def test_get_cert():
> > + """
> > + Test get_cert() function and raises an error in case of UTF8 fixture
> > + """
> > + assert get_cert() is None
> >
> > These should be in test_crypto.py, they're not a test of the debug
> > interface.
> > ------------------------------
> >
> > In tests/test_debug.py
> > <https://github.com/pyca/pyopenssl/pull/1007#discussion_r602599490>:
> >
> > >
> > def test_debug_info():
> > """
> > Debug info contains correct data.
> > """
> > # Just check a sample we control.
> > assert version.__version__ in _env_info
> > +
> > +
> > +def get_cert():
> > + """
> > + Read a certificate, create a x509 object and print his subject
> > + """
> > + x509 = OpenSSL.crypto.load_certificate(
> > + OpenSSL.crypto.FILETYPE_PEM, cert_utf8
> > + )
> > + print(x509.get_subject())
> > +
> > +
> > +def test_get_cert():
> >
> > This isn't correctly named -- the point of this test isn't to test the
> > get_cert function, it's to test non ascii PEM.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/pyca/pyopenssl/pull/1007#pullrequestreview-622550658
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AF7Q4S5DXP5YCMUPNT6T4FTTFT5NFANCNFSM4Z36IYEQ
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pyca/pyopenssl/pull/1007#issuecomment-808620766>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAGBBYXTTPTGSQYBKX3L3TFU3VNANCNFSM4Z36IYEQ>
> .
>


-- 
All that is necessary for evil to succeed is for good people to do nothing.

paulosgf commented:
Okay!

Em sex, 26 de mar de 2021 23:01, Alex Gaynor ***@***.***>
escreveu:

> Please update this PR, don't close and open a new one.
>
> On Fri, Mar 26, 2021 at 9:58 PM Paulo S. G. Ferraz ***@***.***>
> wrote:
>
> > Yes, it's makes sense. I'll change this soon and resubmit.
> >
> > Em sex, 26 de mar de 2021 18:40, Alex Gaynor ***@***.***>
> > escreveu:
> >
> > > ***@***.**** commented on this pull request.
> > > ------------------------------
> > >
> > > In src/OpenSSL/crypto.py
> > > <https://github.com/pyca/pyopenssl/pull/1007#discussion_r602599139>:
> > >
> > > > @@ -1935,7 +1935,7 @@ def load_certificate(type, buffer):
> > > :return: The X509 object
> > > """
> > > if isinstance(buffer, _text_type):
> > > - buffer = buffer.encode("ascii")
> > > + buffer = buffer.encode("ascii", "ignore")
> > >
> > > I don't think ignore makes sense here, can we just encode to utf-8?
> > > ------------------------------
> > >
> > > In tests/test_debug.py
> > > <https://github.com/pyca/pyopenssl/pull/1007#discussion_r602599252>:
> > >
> > > > +
> > > +def get_cert():
> > > + """
> > > + Read a certificate, create a x509 object and print his subject
> > > + """
> > > + x509 = OpenSSL.crypto.load_certificate(
> > > + OpenSSL.crypto.FILETYPE_PEM, cert_utf8
> > > + )
> > > + print(x509.get_subject())
> > > +
> > > +
> > > +def test_get_cert():
> > > + """
> > > + Test get_cert() function and raises an error in case of UTF8 fixture
> > > + """
> > > + assert get_cert() is None
> > >
> > > These should be in test_crypto.py, they're not a test of the debug
> > > interface.
> > > ------------------------------
> > >
> > > In tests/test_debug.py
> > > <https://github.com/pyca/pyopenssl/pull/1007#discussion_r602599490>:
> > >
> > > >
> > > def test_debug_info():
> > > """
> > > Debug info contains correct data.
> > > """
> > > # Just check a sample we control.
> > > assert version.__version__ in _env_info
> > > +
> > > +
> > > +def get_cert():
> > > + """
> > > + Read a certificate, create a x509 object and print his subject
> > > + """
> > > + x509 = OpenSSL.crypto.load_certificate(
> > > + OpenSSL.crypto.FILETYPE_PEM, cert_utf8
> > > + )
> > > + print(x509.get_subject())
> > > +
> > > +
> > > +def test_get_cert():
> > >
> > > This isn't correctly named -- the point of this test isn't to test the
> > > get_cert function, it's to test non ascii PEM.
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <
> https://github.com/pyca/pyopenssl/pull/1007#pullrequestreview-622550658
> > >,
> > > or unsubscribe
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AF7Q4S5DXP5YCMUPNT6T4FTTFT5NFANCNFSM4Z36IYEQ
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/pyca/pyopenssl/pull/1007#issuecomment-808620766>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAAAGBBYXTTPTGSQYBKX3L3TFU3VNANCNFSM4Z36IYEQ
> >
> > .
> >
>
>
> --
> All that is necessary for evil to succeed is for good people to do nothing.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pyca/pyopenssl/pull/1007#issuecomment-808621250>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF7Q4S54KXRLFL6Y6TYQ3HDTFU365ANCNFSM4Z36IYEQ>
> .
>

paulosgf commented:
Changed OpenSSL.crypto.load_certificate to use buffer.encode("utf-8") instead ("ascii', "ignore").
Moved tests to tests/test_crypto.utf8_pem() and test_utf8_pem()

alex commented on CHANGELOG.rst:
```diff
@@ -19,6 +19,8 @@ Deprecations:
 Changes:
 ^^^^^^^^
 
+- Fixed exception raised when load an certificate with UTF-8 characters on ``Openssl.crypto.load_certificate``
```
```suggestion
- Fixed exception raised when loading a certificate with UTF-8 characters in ``Openssl.crypto.load_certificate``
```

alex commented on tests/test_crypto.py:
```diff
@@ -4454,3 +4484,20 @@ def test_does_not_contain(self):
         assert curve not in curves
+
+
+def utf8_pem():
+    """
+    Read a certificate, create a x509 object and print his subject
+    """
+    x509 = OpenSSL.crypto.load_certificate(
+        OpenSSL.crypto.FILETYPE_PEM, cert_utf8
+    )
+    print(x509.get_subject())
+
+
+def test_utf8_pem():
+    """
+    Test utf8_pemt() function and raises an error in case of UTF8 fixture
+    """
+    assert utf8_pem() is None
```
You don't need a seperate `utf8_pem` function, and please do not print the certificate. Also, please assert something more specific about the result.

paulosgf reviewed (commented):
Hi guys!
Someone could review this bugfix for me? Please!

reaperhulk commented on tests/test_crypto.py:
```diff
@@ -4454,3 +4484,12 @@ def test_does_not_contain(self):
+
+def test_utf8_pem():
+    """
+    Read a certificate and raises an error in case of UTF8 fixture.
```
This comment is incorrect.

paulosgf replied:
Hi reaperhulk! 

Maybe something like this?

Checks if an UTF-8 certificate raises an encoding exception.

reaperhulk replied:
"Checks that a PEM encoded file with UTF-8 characters outside the PEM boundary does not raise an exception"

paulosgf reviewed (commented):
@reaperhulk, can you review the changes requested?
